trin-core/src/portalnet/overlay_service.rs

```diff
@@ -643,27 +643,42 @@ impl<TContentKey: OverlayContentKey + Send, TMetric: Metric + Send>
         }
     }
 
-    /// Attempts to build a `Accept` response for a `Offer` request.
+    /// Attempts to build a `Accept` response for an `Offer` request.
     fn handle_offer(&self, request: Offer) -> Result<Accept, OverlayRequestError> {
         self.metrics
             .as_ref()
             .and_then(|m| Some(m.report_inbound_offer()));
-        let mut requested_keys = BitList::with_capacity(request.content_keys.len())
-            .map_err(|e| OverlayRequestError::AcceptError(e))?;
-        let conn_id: u16 = crate::utp::stream::rand();
 
-        // TODO: Pipe this with overlay DB and request only not available keys.
+        let mut requested_keys =
+            BitList::with_capacity(request.content_keys.len()).map_err(|_| {
+                OverlayRequestError::AcceptError(
+                    "Unable to initialize bitlist for requested keys.".to_owned(),
+                )
+            })?;
+
         let accept_keys = request.content_keys.clone();
 
-        for (i, key) in request.content_keys.iter().enumerate() {
-            // should_store is currently a dummy function
-            // the actual function will take ContentKey type, so we'll  have to decode keys here
+        for (i, key) in request.content_keys.into_iter().enumerate() {
+            // TODO: This will not work when we receive state Offer message with state content keys
+            // we need to refactor it in a way to get access to the protocol id in this method
+            let key = HistoryContentKey::try_from(key)
+                .map_err(|err| OverlayRequestError::AcceptError(err.to_string()))?;
```

Comment:
I think that instead of having to check the protocol ID, we can use the generic `TContentKey` to do this:

```suggestion
           let key = (TContentKey::try_from)(key)
                .map_err(|err| OverlayRequestError::AcceptError("Invalid Content Key".to_string()))?;
```

This is also done [here](https://github.com/ethereum/trin/blob/26deb62b3a9e00af308ab0bf0f6c20b81f3d1153/trin-core/src/portalnet/overlay_service.rs#L589)

Reply:
Nice, this is exactly what we need :100:!